File: 백준/Gold/1939. 중량제한/중량제한.java
import java.io.*;
import java.util.*;
public class Main {
  
  static int N, M;
  static int[] P;
  static List<int[]> Edges = new ArrayList<>();

  public static int swap(int x, int y){ return x; }

  public static int find(int x){
    if(P[x] < 0) return x;
    return P[x] = find(P[x]);
  }
  
  public static void union(int x, int y){
    x = find(x);
    y = find(y);
    if(x == y) return;
    if(P[x] == P[y]) P[x]--;
    if(P[x] > P[y]) x = swap(y, y = x);
    P[y] = x;
  }

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();
    StringTokenizer st = new StringTokenizer(br.readLine());

    // input
    N = Integer.parseInt(st.nextToken());
    M = Integer.parseInt(st.nextToken());
        
    // 간선 정보
    for(int i = 0; i < M; i++){
      st = new StringTokenizer(br.readLine());
      int u = Integer.parseInt(st.nextToken());
      int v = Integer.parseInt(st.nextToken());
      int w = Integer.parseInt(st.nextToken());
      Edges.add(new int[]{w, u, v});
    }

    // 출발지와 목적지 입력 받기
    st = new StringTokenizer(br.readLine());
    int start = Integer.parseInt(st.nextToken());
    int end = Integer.parseInt(st.nextToken());

    // 간선 정보를 가중치를 기준으로 내림차순 정렬
    Edges.sort(new Comparator<int[]>() {
      @Override
      public int compare(int[] e1, int[] e2){
        return Integer.compare(e2[0], e1[0]);
      }
    });

    // init
    P = new int[N+1];
    for(int i = 0; i <= N; i++) P[i] = -1;

    // solve
    for(int[] cur : Edges){
      union(cur[1], cur[2]);
      if(find(start) == find(end)){
        sb.append(cur[0]);
        break;
      }
    }

    // output
    System.out.println(sb.toString());
  }  
}

File: 백준/Gold/1939. 중량제한/중량제한.py
#Authored by : chjh2129
import sys
'''
  간선 정보를 가중치가 큰 순서부터 순차적으로 merge를 진행하여 풀이합니다.

  임의의 간선 (w, u, v)가 merge가 되고 출발지(st)와 목적지(en)가 같은 집합에 속하게되었다면.

  이전에 연결된 간선의 가중치는 w보다 크거나 같습니다. 이후에 연결할 간선 정보는 w보다 작거나 같으므로,

  출발지에서 목적지로 들고갈 수 있는 중량의 최댓값은 w가 됩니다.
'''
input = sys.stdin.readline

def find(x: int, P: list[int]) -> int:
  if P[x] < 0: return x
  P[x] = find(P[x], P)
  return P[x]

def union(x: int, y: int, P: list[int]) -> None:
  x = find(x, P)
  y = find(y, P)
  if x == y: return
  if P[x] == P[y]: P[x] -= 1
  if P[x] > P[y]: x, y = y, x
  P[y] = x

def main() -> None:
  # 섬 개수, 다리 개수
  N, M = map(int, input().split())
  
  # 간선 정보 저장
  E = [ ( w, u, v ) for u, v, w in (map(int, input().split()) for _ in range(M)) ]

  # 출발지와 목적지
  st, en = map(int, input().split())
  
  # init 
  P: list[int] = [ -1 for _ in range(N+1) ]
  E.sort()

  # solve
  while E:
    w, u, v = E.pop()
    union(u, v, P)
    if(find(st, P) == find(en, P)):
      # 출력 후 종료
      print(w)
      break

    
sys.exit(main())